### ISSUE TITLE
Publish messages out-of-order without waiting the result

### ISSUE THREAD

**Issue by vwxyzh** [StackExchange/StackExchange.Redis#1719]
[Defect code](https://github.com/StackExchange/StackExchange.Redis/blob/main/src/StackExchange.Redis/PhysicalBridge.cs#L959):
```cs
token = _singleWriterMutex.TryWait(options: WaitOptions.NoDelay);
if (!token.Success)
{
   if (PushToBacklog(message, onlyIfExists: !ALWAYS_USE_BACKLOG_IF_CANNOT_GET_SYNC_LOCK))
      return new ValueTask<WriteResult>(WriteResult.Success); // queued counts as success
   ...
}
...
if (!flush.IsCompletedSuccessfully)
{
  ...
  return CompleteWriteAndReleaseLockAsync(token, flush, message, lockTaken);
}
```

Scenario:
Thread A:
```cs
await subscriber.PublishAsync("channel-1", 1);
```
Thread B:
```cs
await Task.WhenAll(subscriber.PublishAsync("channel-2", 2), subscriber.PublishAsync("channel-2", 3));
```
1. Thread A try to publish message `1`
2. Thread A takes the lock(_singleWriterMutex) first.
3. Thread B try to publish message `2`
4. Thread B cannot take the lock, send to backlog.
5. Thread B create the backlog, and start the background processing.
6. Thread B returns, since `queued counts as success`.
7. Thread A finish the write message `1`, release the lock.
8. Thread B try to publish message `3`
9. Thread B takes the lock.
10. Thread B finish the write message `3`, release the lock.
11. The backlog takes the lock.
12. The backlog write message `2`.

Then, finally we get following result:
Channel-1: `1`
Channel-2: `3, 2`

**Comment by TimLovellSmith** [comment]
@vwxyzh If you want the messages to arrive in order, you would need to send them in order, so why are you doing 

```
await Task.WhenAll(subscriber.PublishAsync("channel-2", 2), subscriber.PublishAsync("channel-2", 3));
```
Which has no ordering guarantee on task execution, since tasks get distributed across thread pool threads?

Shouldn't you instead do
```
await subscriber.PublishAsync("channel-2", 2);
await subscriber.PublishAsync("channel-2", 3);
```
?

**Comment by vwxyzh** [comment]
@TimLovellSmith that `await` will wait for redis server ack, it will be slower than expected.
And it will keep order when client in idle mode (do not use the backlog) or in busy mode (all messages queued in backlog),
but it breaks the order when the client switch it from idle mode to busy mode.
So i think it is a bug.

**Comment by TimLovellSmith** [comment]
Oh, thanks for the clearer description. Then in that case, I think this the same issue as https://github.com/StackExchange/StackExchange.Redis/issues/1440

**Comment by TimLovellSmith** [comment]
Some ideas for fixes:

1. Have incoming calls to WriteMessageTakingWriteLockSync/Async always write to the backlog, if there is a backlog currently, until the backlog is cleared by the backlog processor. That will avoid their message jumping the backlog queue, and getting sent out of order

2. Have incoming calls to WriteMessageTakingWriteLockSync/Async attempt to clear the backlog first, if there is a small enough backlog currently, now that they have already acquired the lock. They will have to clear the whole backlog before sending the latest message. If unable to clear the backlog, they'll have to leave their message in the backlog and back.

**Comment by mgravell** [comment]
Would the CreateBatch or CreateTransaction APIs help here?

**Comment by TimLovellSmith** [comment]
@mgravell CreateBatch for workarounds, you mean?

Any thoughts on the fix ideas? I am thinking about doing a PR on this but I am a little worried about this code being in the hot path a lot, so I'm interested in if you have any concern with predictable perf issues with those approaches.

**Comment by TimLovellSmith** [comment]
I've experimented with the first fix idea, which is to atomically check and either add to the backlog, or try to take the write lock if the backlog is empty. Unfortunately it causes many tests to fail. 

The majority of them are failing with ConnectTimeout which was a surprise to me. I would intuit that when we add to the backlog more aggressively, it is likely to increase the rate at which WriteMessageTakingWriteLockAsync return WriteResult.Success in an optimistic "yup, queued the message!" sense.

But, it could also in general, take longer for those message to get asynchronously dequeued and processed. Perhaps that is the problem. Or maybe my troubles stem from contention...

>[xUnit.net 00:00:12.21]     ConnectFailTimeout.NoticesConnectFail [FAIL]
  Failed ConnectFailTimeout.NoticesConnectFail [10 s]
  Error Message:
   System.TimeoutException : Connect timeout
  Stack Trace:
     at StackExchange.Redis.Tests.TestBase.CreateDefault(TextWriter output, String clientName, Nullable`1 syncTimeout, Nullable`1 allowAdmin, Nullable`1 keepAlive, Nullable`1 connectTimeout, String password, String tieBreaker, TextWriter log, Boolean fail, String[] disabledCommands, String[] enabledCommands, Boolean checkConnect, String failMessage, String channelPrefix, Nullable`1 proxy, String configuration, Boolean logTransactionData, Nullable`1 defaultDatabase, String caller) in C:\repo\StackExchange.Redis\tests\StackExchange.Redis.Tests\TestBase.cs:line 354
   at StackExchange.Redis.Tests.TestBase.Create(String clientName, Nullable`1 syncTimeout, Nullable`1 allowAdmin, Nullable`1 keepAlive, Nullable`1 connectTimeout, String password, String tieBreaker, TextWriter log, Boolean fail, String[] disabledCommands, String[] enabledCommands, Boolean checkConnect, String failMessage, String channelPrefix, Nullable`1 proxy, String configuration, Boolean logTransactionData, Boolean shared, Nullable`1 defaultDatabase, String caller) in C:\repo\StackExchange.Redis\tests\StackExchange.Redis.Tests\TestBase.cs:line 230
   at StackExchange.Redis.Tests.ConnectFailTimeout.<NoticesConnectFail>d__1.MoveNext() in C:\repo\StackExchange.Redis\tests\StackExchange.Redis.Tests\ConnectFailTimeout.cs:line 16
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
  Standard Output Messages:
 Running under .NET Framework 4.8.4341.0 (4.0.30319.42000)
   Running on: Microsoft Windows 10.0.19042
 21:40:17.188: 127.0.0.1:6379,name=NoticesConnectFail,allowAdmin=True
 Connecting 127.0.0.1:6379/Interactive...
 BeginConnect: 127.0.0.1:6379
 1 unique nodes specified
 Requesting tie-break from 127.0.0.1:6379 > __Booksleeve_TieBreak...
 Allowing endpoints 00:00:05 to respond...
 Awaiting task completion, IOCP: (Busy=8,Free=992,Min=8,Max=1000), WORKER: (Busy=8,Free=2039,Min=8,Max=2047)
 Connected Interactive/127.0.0.1:6379
 Server handshake
 Setting client name: NoticesConnectFail
 Auto-configure...
 Sending critical tracer: Interactive/127.0.0.1:6379
 Flushing outbound buffer
 Starting read
 Not all tasks completed cleanly (from ReconfigureAsync#1748, timeout 5000ms), IOCP: (Busy=0,Free=1000,Min=8,Max=1000), WORKER: (Busy=8,Free=2039,Min=8,Max=2047)
 127.0.0.1:6379 did not respond
 Waiting for tiebreakers...
 Awaiting task completion, IOCP: (Busy=0,Free=1000,Min=8,Max=1000), WORKER: (Busy=8,Free=2039,Min=8,Max=2047)
 Not all tasks completed cleanly (from NominatePreferredMaster#1985, timeout 50ms), IOCP: (Busy=0,Free=1000,Min=8,Max=1000), WORKER: (Busy=8,Free=2039,Min=8,Max=2047)
 127.0.0.1:6379 failed to nominate (WaitingForActivation)
 No masters detected
 127.0.0.1:6379: Standalone v3.0.503, master; 2000 databases; keep-alive: 00:01:00; int: ConnectedEstablishing; sub: ConnectedEstablished, 1 active; not in use: DidNotRespond
 127.0.0.1:6379: int ops=6, qu=0, qs=0, qc=0, wr=0, socks=1; sub ops=3, qu=0, qs=0, qc=0, wr=0, subs=1, socks=1
 Circular op-count snapshot; int: 0+6=6 (0.60 ops/s; spans 10s); sub: 0+3=3 (0.30 ops/s; spans 10s)
 Sync timeouts: 0; async timeouts: 0; fire and forget: 0; last heartbeat: -1s ago
 resetting failing connections to retry...
 retrying; attempts left: 2...
 1 unique nodes specified
 Requesting tie-break from 127.0.0.1:6379 > __Booksleeve_TieBreak...
 Allowing endpoints 00:00:05 to respond...
 Awaiting task completion, IOCP: (Busy=1,Free=999,Min=8,Max=1000), WORKER: (Busy=8,Free=2039,Min=8,Max=2047)
 connection failed: 127.0.0.1:6379 (Interactive, UnableToConnect): UnableToConnect (None, last-recv: 2148) on 127.0.0.1:6379/Interactive, Flushed/ReadAsync, last: CLUSTER, origin: ResetNonConnected, outstanding: 0, last-read: 4s ago, last-write: 4s ago, keep-alive: 60s, state: ConnectedEstablishing, mgr: 9 of 10 available, last-heartbeat: never, global: 5s ago, v: 2.2.22.6678
 21:40:17.189: Service Counts: (Scheduler) Queue: 1618, Pool: 0

**Comment by TimLovellSmith** [comment]
Actually now I look more closely, seems likely the main problem is just that its running me out of free worker threads.
> Awaiting task completion, IOCP: (Busy=0,Free=1000,Min=8,Max=1000), WORKER: (Busy=7,Free=2040,Min=8,Max=2047)

**Comment by TimLovellSmith** [comment]
Should be fixed in next release via e8808355ccf3540983a82074a49cfae425dd8d97.

**Comment by TimLovellSmith** [comment]
Breadcrumb remark! For anyone lost in the woods, and still seeing issues in 'NominatePreferredMaster' with the 50ms timeout, like mentioned in passing here, I'd strongly suggest you update to a more recent version of the library (such as latest and greatest, 2.6.48) to take advantage of many fixes to connection reliability that relate to that code path, and _not just any version newer than this discussion_. Cheers!
